crates/engine/src/rocksdb_engine/transaction.rs

```diff
@@ -1,18 +1,231 @@
 #![allow(clippy::module_name_repetitions)]
-#![allow(clippy::multiple_inherent_impl)]
 
-use crate::TransactionApi;
+use std::{
+    iter::repeat,
+    sync::{
+        atomic::{AtomicU64, AtomicUsize, Ordering},
+        Arc,
+    },
+};
 
-/// A transaction of the `RocksEngine`
-#[derive(Copy, Clone, Debug, Default)]
-pub struct RocksTransaction;
+use clippy_utilities::NumericCast;
+use parking_lot::Mutex;
+use rocksdb::{Direction, IteratorMode, OptimisticTransactionDB, Transaction};
+
+use crate::{api::transaction_api::TransactionApi, error::EngineError, StorageOps, WriteOperation};
+
+use super::RocksEngine;
+
+/// Transaction type for `RocksDB`
+pub struct RocksTransaction<'db> {
+    /// The inner DB
+    db: Arc<OptimisticTransactionDB>,
+    /// A transaction of the DB
+    ///
+    /// We need a `Mutex` because `Transaction<'db, DB>` does not implement `Sync`
+    txn: Mutex<Option<Transaction<'db, OptimisticTransactionDB>>>,
+    /// The size of the engine
+    engine_size: Arc<AtomicU64>,
```

Comment:
The `AtomicU64` has already implemented `Sync` and `Send`. Why should we wrap it in a `Arc`? Can we use `&'db AtomicU64` to replace it?

Reply:
done